src/relay.cpp

```diff
@@ -593,20 +630,71 @@ void relay_relay_forw(int sock, const uint8_t *msg, int32_t len, const ip6_hdr *
  *
  * @return              none
  */
-void callback(evutil_socket_t fd, short event, void *arg) {
-    struct relay_config *config = (struct relay_config *)arg;
+void client_callback(evutil_socket_t fd, short event, void *arg) {
+    auto vlans = *(reinterpret_cast<std::unordered_map<std::string, struct relay_config> *>(arg));
     static uint8_t message_buffer[BUFFER_SIZE];
-    std::string counterVlan = counter_table;
-    int32_t len = recv(config->filter, message_buffer, BUFFER_SIZE, 0);
-    if (len <= 0) {
-        syslog(LOG_WARNING, "recv: Failed to receive data at filter socket: %s\n", strerror(errno));
-        return;
+    struct sockaddr_ll sll;
+    socklen_t slen = sizeof(sll);
+    int pkts_num = 0;
+
+    while (pkts_num++ < BATCH_SIZE) {
+        auto buffer_sz = recvfrom(fd, message_buffer, BUFFER_SIZE, 0, (struct sockaddr *)&sll, &slen);
```

Comment:
Just Curious, Does the sll be populated by the ifindex of Vlan netdev even when the client is an untagged member?

Reply:
The filter raw socket doesn't bind to particular interface, so raw socket will filter all system interfaces.
Including vlan member interface, Bridge, Vlan, lo ...

For example when Solicit comes from untagged member, first member interface will has this packet as input, then Bridge has this packet as input, and then vlan interface will has also has this packet as input. 
In our logic we will receive 3 copy of same Solicit packet, and sll has the related input interface info. And in our recvfrom logic we only care about vlan member interfaces, other 2 duplicate packets from Bridge and Vlan are ignored. 

Another simple solution maybe we just bind filter raw sockets to all untagged members, but with this way we would not be easy to handle interface (state and vlan member) dynamic change and also will generate more sockets. And also we can't bind only on vlan interfaces, as we need the physical interface info for dual-tor case and in future for the purpose of interface counter statistics support. 

```
struct sockaddr_ll sll = {
    .sll_family = AF_PACKET,
    .sll_protocol = htons(ETH_P_ALL),
    .sll_ifindex = 0 // any interface
};
```

Reply:
Thanks for the explanation, i get the flow. Binding to all ifaces and filtering the unwanted might be the best implementation here.